pallets/pallet-bonded-coins/src/lib.rs

```diff
@@ -199,32 +214,195 @@ pub mod pallet {
+					.unwrap_or(Zero::zero());
+
+				if !amount.is_zero() {
+					T::CollateralCurrency::transfer(
+						T::CollateralAssetId::get(),
+						&pool_account,
+						&who,
+						amount,
+						Preservation::Expendable,
+					)?;
+					total_collateral_issuance.saturating_reduce(amount);
+				}
+
+				sum_of_issuances.saturating_reduce(burnt);
+
+				// if the total issuance drops to 0 due to this burn, kill the currency
+				if token_issuance <= &burnt {
+					dead_currencies.push(currency_id);
+				}
+			}
+
+			// destroy all active currencies if the collateral locked in the pool has dropped to 0
+			if total_collateral_issuance.is_zero() || sum_of_issuances.is_zero() {
+				// we assume all currencies that already had their total issuance at 0 have already been deactivated in a previous step
+				for (currency_id, _) in total_issuances {
+					T::Fungibles::start_destroy(currency_id, None)?;
+				}
+				Self::deposit_event(Event::RefundCompleted(pool_id.clone()));
+			} else {
+				// deactivate all currencies whose issuance dropped to 0 in this step
+				for currency_id in dead_currencies {
+					T::Fungibles::start_destroy(currency_id.clone(), None)?;
+				}
+			}
+
+			Ok(())
 		}
 
 		#[pallet::call_index(9)]
 		#[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))]
-		pub fn finish_destroy(_origin: OriginFor<T>) -> DispatchResult {
-			todo!()
+		pub fn finish_destroy(origin: OriginFor<T>, pool_id: T::PoolId) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			let pool_details = Pools::<T>::get(pool_id.clone()).ok_or(Error::<T>::PoolUnknown)?;
+
+			ensure!(pool_details.state.is_destroying(), Error::<T>::InUse);
+
+			for currency_id in pool_details.bonded_currencies {
+				if T::Fungibles::asset_exists(currency_id.clone()) {
+					// This would fail with an InUse error if there are any accounts left on any currency
+					T::Fungibles::finish_destroy(currency_id.clone())?;
+				}
+			}
+
+			let pool_account = pool_id.clone().into();
+
+			let total_collateral_issuance = Self::get_pool_collateral(&pool_account);
+
+			if !total_collateral_issuance.is_zero() {
+				T::CollateralCurrency::transfer(
+					T::CollateralAssetId::get(),
+					&pool_account,
+					&pool_details.manager,
+					total_collateral_issuance,
+					Preservation::Expendable,
+				)?;
+			}
+
+			Pools::<T>::remove(&pool_id);
+
+			// TODO: refund deposit
+
+			Self::deposit_event(Event::Destroyed(pool_id));
+
+			Ok(())
 		}
 	}
 
-	impl<T: Config> Pallet<T> {}
+	impl<T: Config> Pallet<T> {
+		fn do_start_destroy_pool(pool_id: T::PoolId, maybe_check_owner: Option<AccountIdOf<T>>) -> DispatchResult {
+			let mut pool_details = Pools::<T>::get(pool_id.clone()).ok_or(Error::<T>::PoolUnknown)?;
+
+			ensure!(!pool_details.state.is_destroying(), Error::<T>::Destroying);
+
+			if let Some(caller) = maybe_check_owner {
+				ensure!(pool_details.is_manager(&caller), Error::<T>::Unauthorized);
+			}
+
+			pool_details.state.destroy();
```

Comment:
Hence I would probably not make `pool_details` mut, but simply call `Pools::<T>::mutate` later, which gives you a mutable reference to the pool, and in there I would call `.destroy()`. You can check [here](https://github.com/KILTprotocol/kilt-node/blob/f057961868d2d865a8d3ff4a5f977cc5e381472a/pallets/pallet-asset-switch/src/xcm/transact/mod.rs#L111) for an example.

Reply:
Thanks, I missed that! bb2c1c4f112082c02963f4d99f4b834046c37d84

Reply:
Ok this is already better. I'm not a big fan of having unrelated operations inside the storage entry stuff, and also having a read and later a write won't impact on performance since storage keys are anyway written to an overlay before being written into storage, so it would not make performance any worse to have a get first, and a mutate later.

Reply:
I see, I've based this on patterns and implementations I've seen in the assets pallet, where things are usually handled this way. But I have no strong preference here.

Reply:
Combining a `::get()` with a `::try_mutate()` seems odd to me though, so I went for a `::set()` instead. Why would we look up twice, including duplicated checks if the entry even exists, and the necessary error handling? Isn't this weird?

Reply:
d58885b04ad7edd0f0f59639a5720a39720a9a35

Reply:
Yes, `set()` works as well.